### PR TITLE
Support multiple proof purposes in `verify`.

### PR DESCRIPTION
**Pull request by dlongley**
Includes cache of document c14n hash to avoid duplicate work when verifying multiple proofs against the same document.

### CODE REVIEW

**Comment by dlongley** [comment]
This has been tested against zcapld and ed25519-signature-2020 using their separate test suites. We should really have a test suite here as well.

File: lib/ProofSet.js [hunk 1]
```diff
@@ -146,12 +146,14 @@ module.exports = class ProofSet {
 
       // verify proofs
       const results = await _verify({
-        document, suites, proofSet, purpose, documentLoader, expansionMap
+        document, suites, proofSet, purpose, documentLoader, expansionMap,
```
Comment (msporny) [inline lib/ProofSet.js:149]:
```suggestion
        document, suites, proofSet, purpose, documentLoader, expansionMap
```

Remove hanging comma.

File: lib/ProofSet.js [hunk 2]
```diff
@@ -197,38 +199,108 @@ async function _getProofs({document}) {
...
+      };
+      const {verified, verificationMethod, error} = await suite.verifyProof({
+        proof, document, purpose, documentLoader, expansionMap
+      });
+      if(!vm) {
+        vm = verificationMethod;
       }
+      result = {proof, verified, verificationMethod: vm, error};
+    } catch(error) {
+      result = {proof, verified: false, error};
```
Comment (msporny) [inline lib/ProofSet.js:240]: Do we need a `_makeSerializable(error)` here?

File: lib/ProofSet.js [hunk 3]
```diff
@@ -197,38 +199,108 @@ async function _getProofs({document}) {
...
+        // validate purpose
+        const {verificationMethod} = result;
+        const suite = proofToSuite.get(proof);
+        let purposeResult;
+        try {
+          purposeResult = await purpose.validate(proof, {
+            document, suite, verificationMethod, documentLoader, expansionMap
+          });
+        } catch(error) {
+          purposeResult = {valid: false, error};
```
Comment (msporny) [inline lib/ProofSet.js:270]: Again, do we need a `_makeSerializable(error)` call here?

File: lib/ProofSet.js [hunk 4]
```diff
@@ -237,3 +309,52 @@ function _addToJSON(error) {
...
+  suiteMatchQueue, document, documentLoader, expansionMap
+}) {
+  for(const proof of proofSet) {
+    // first check if the proof matches the purpose; if it doesn't continue
+    if(!await purpose.match(proof, {document, documentLoader, expansionMap})) {
+      continue;
+    }
+
+    // next, find the suite that can verify the proof
+    let matched = false;
```
Comment (msporny) [inline lib/ProofSet.js:324]: It's not clear what happens if `matched` continues to be `false`? I'm guessing `matches` is processed in some way in the calling function... hard to follow this code.

**Review by msporny — CHANGES_REQUESTED** [review]
I'd like to see a test that tests the multiple proofs path, do we have one? If not, I suggest adding one before merging. I expect it would be easy for another developer to break multi-proof verification and not know they did without that test.

**Comment by dlongley** [comment]
The multiple proofs test is over here:

https://github.com/digitalbazaar/zcapld/pull/64/files#diff-dce1aec60ad1f68ca409661764b88da658d166287e3e107738f901d0052be2e6R535

File: lib/ProofSet.js (hunk 2 above)
Comment (dlongley) [reply lib/ProofSet.js:240]: No, it happens below in one place: https://github.com/digitalbazaar/jsonld-signatures/pull/149/files#diff-01c3bd2f5d6eee0565c498c5535384249141782a41d6b8d5da84a6fc183a96e7R245

File: lib/ProofSet.js (hunk 3 above)
Comment (dlongley) [reply lib/ProofSet.js:270]: No, it happens below in one place: https://github.com/digitalbazaar/jsonld-signatures/pull/149/files#diff-01c3bd2f5d6eee0565c498c5535384249141782a41d6b8d5da84a6fc183a96e7R287

File: lib/ProofSet.js (hunk 4 above)
Comment (dlongley) [reply lib/ProofSet.js:324]: I'll add a comment about it.

Comment (dlongley) [reply lib/ProofSet.js:327]: https://github.com/digitalbazaar/jsonld-signatures/pull/149/commits/bc539b6170398a2f2e7e791a9991b5648ef096ab

File: lib/ProofSet.js (hunk 1 above)
Comment (dlongley) [reply lib/ProofSet.js]: Removed.

**Review by msporny — APPROVED** [review]
